core/src/main/java/org/apache/brooklyn/core/catalog/internal/BundleUpgradeParser.java

```diff
@@ -0,0 +1,239 @@
+                return this;
+            }
+            public CatalogUpgrades build() {
+                return new CatalogUpgrades(this);
+            }
+        }
+        
+        public static Builder builder() {
+            return new Builder();
+        }
+        
+        private Set<VersionRangedName> removedLegacyItems;
+        private Set<VersionRangedName> removedBundles;
+        
+        public CatalogUpgrades(Builder builder) {
+            this.removedLegacyItems = ImmutableSet.copyOf(builder.removedLegacyItems);
+            this.removedBundles = ImmutableSet.copyOf(builder.removedBundles);
+        }
+
+        public boolean isEmpty() {
+            return removedLegacyItems.isEmpty() && removedBundles.isEmpty();
+        }
+
+        public Set<VersionRangedName> getRemovedLegacyItems() {
+            return removedLegacyItems;
+        }
+        
+        public Set<VersionRangedName> getRemovedBundles() {
+            return removedBundles;
+        }
+
+        public boolean isLegacyItemRemoved(CatalogItem<?, ?> legacyCatalogItem) {
+            VersionedName name = new VersionedName(legacyCatalogItem.getSymbolicName(), legacyCatalogItem.getVersion());
+            return contains(removedLegacyItems, name);
+        }
+
+        public boolean isBundleRemoved(VersionedName bundle) {
+            return contains(removedBundles, bundle);
+        }
+        
+        public boolean contains(Iterable<VersionRangedName> names, VersionedName name) {
+            for (VersionRangedName contender : names) {
+                if (contender.getSymbolicName().equals(name.getSymbolicName()) && contender.getOsgiVersionRange().includes(name.getOsgiVersion())) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+    
+    /**
+     * Records a name (string) and version range (string),
+     * with conveniences for pretty-printing and converting to OSGi format.
+     * 
+     * Implementation-wise, this is similar to {@link VersionedName}, but is intended
+     * as internal-only so is cut down to only what is needed.
+     */
+    public static class VersionRangedName {
+        private final String name;
+        private final String v;
+        private transient volatile VersionRange cachedOsgiVersionRange;
+
+        public static VersionRangedName fromString(String val, boolean singleVersionIsOsgiRange) {
+            if (Strings.isBlank(val)) {
+                throw new IllegalArgumentException("Must not be blank");
+            }
+            String[] parts = val.split(":");
+            if (parts.length > 2) {
+                throw new IllegalArgumentException("Identifier '"+val+"' has too many parts; max one ':' symbol");
+            }
+            if (parts.length == 1 || Strings.isBlank(parts[1])) {
+                throw new IllegalArgumentException("Identifier '"+val+"' must be of 'name:versionRange' syntax");
```

Comment:
worth noting that range is required by this parser, and in docs we need to specify where ranges are required and where they are optional